openwisp_firmware_upgrader/upgraders/openwrt.py

```diff
@@ -195,14 +197,34 @@ def _verify_device_uuid(self):
 
     def upgrade(self, image):
         self._test_connection()
+        self._check_cancellation()
         self._verify_device_uuid()
+        self._check_cancellation()
         checksum = self._test_checksum(image)
+        self._check_cancellation()
         remote_path = self.get_remote_path(image)
-        self.upload(image.file, remote_path)
+        self.upload(image, remote_path)
+        self._check_cancellation()
         self._test_image(remote_path)
         self._reflash(remote_path)
         self._write_checksum(checksum)
 
+    def _check_cancellation(self):
+        """
+        Check if the upgrade operation has been cancelled.
+        """
+        self.upgrade_operation.refresh_from_db()
+
+        if self.upgrade_operation.status == "cancelled":
+            self.log(_("Upgrade operation has been cancelled by user"))
```

Comment:
Hasn't this already been done by the `cancel` method in the upgrade operation model? Seems redundant, please remove if redundancy is confirmed.